function-sidecar/pkg/dispatcher/grpc/grpc_test.go

```diff
@@ -17,30 +17,199 @@
+	Context("with size based strategy messages", func() {
+		BeforeEach(func() {
+			strategy = v1alpha1.Windowing{Size: 2}
+		})
+		It("should use different streams", func() {
+			go func() {
+				GinkgoRecover()
+				d.Input() <- message.NewMessage([]byte("1"), nil)
+				d.Input() <- message.NewMessage([]byte("2"), nil)
+				d.Input() <- message.NewMessage([]byte("5"), nil)
+				d.Input() <- message.NewMessage([]byte("4"), nil)
+			}()
+
+			received := make(map[string]struct{})
+			m := <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+
+			expected := map[string]struct{}{"1": {}, "3": {}, "5": {}, "9": {}}
+
+			Expect(received).To(Equal(expected))
+		})
+	})
+
+	Context("with time based strategy messages", func() {
+		BeforeEach(func() {
+			strategy = v1alpha1.Windowing{Time: "10ms"}
+		})
+		It("should use different streams", func() {
+			go func() {
+				GinkgoRecover()
+				d.Input() <- message.NewMessage([]byte("1"), nil)
+				d.Input() <- message.NewMessage([]byte("2"), nil)
+				time.Sleep(15 * time.Millisecond)
+				d.Input() <- message.NewMessage([]byte("5"), nil)
+				d.Input() <- message.NewMessage([]byte("4"), nil)
+			}()
+
+			received := make(map[string]struct{})
+			m := <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+
+			expected := map[string]struct{}{"1": {}, "3": {}, "5": {}, "9": {}}
+
+			Expect(received).To(Equal(expected))
+		})
+	})
+	Context("with session based strategy messages", func() {
+		BeforeEach(func() {
+			strategy = v1alpha1.Windowing{Session: "20ms"}
+		})
+		It("should use different streams", func() {
+			go func() {
+				GinkgoRecover()
+				d.Input() <- message.NewMessage([]byte("100"), nil)
+				time.Sleep(5 * time.Millisecond)
+				d.Input() <- message.NewMessage([]byte("50"), nil)
+				time.Sleep(5 * time.Millisecond)
+				d.Input() <- message.NewMessage([]byte("20"), nil)
+				time.Sleep(30 * time.Millisecond)
+				d.Input() <- message.NewMessage([]byte("54"), nil)
+				time.Sleep(5 * time.Millisecond)
+				d.Input() <- message.NewMessage([]byte("44"), nil)
+			}()
+
+			received := make(map[string]struct{})
+			m := <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+			m = <-d.Output()
+			received[string(m.Payload())] = struct{}{}
+
+			expected := map[string]struct{}{"100": {}, "150": {}, "170": {}, "54": {}, "98": {}}
+
+			Expect(received).To(Equal(expected))
```

Comment:
Why is this doing a set-wise comparison since the ordering seems to be important?

Reply:
The ordering *is* indeed important, but not that much. What's more important to notice is that results of 2 streams can be received in parallel, so I chose a setup that can cope with that easily at the detriment of order checking (per stream). Ordering is checked in first tests for what it's worth